### PR TITLE
gcm clients now have separate instances.

### PR DESCRIPTION
**Pull request by bolshoy**
Each instance contains internal http and xmpp client.
Xmpp client runs ping check and tries to reconnect after ping failure.

### CODE REVIEW

**Comment by oryband** [comment]
delete the merged branches

File: gcm.go [hunk 1]
```diff
@@ -158,375 +58,81 @@ type Notification struct {
...
-}
-
-// httpGcmClient is a client for the Gcm Http Connection Server.
-type httpGcmClient struct {
-	GcmURL     string
-	HttpClient *http.Client
-	retryAfter string
+type Client struct {
+	senderID   string
+	apiKey     string
```
Comment (oryband) [inline gcm.go]:
```go
senderID,
apiKey string
```

File: http_client.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+func (c *httpGcmClient) sendHttp(m HttpMessage, b backoffProvider) (*HttpResponse, error) {
+	// TODO(silvano): check this with responses for topic/notification group
+	gcmResp := &HttpResponse{}
+	var multicastId int
+	targets, err := messageTargetAsStringsArray(m)
+	if err != nil {
+		return gcmResp, fmt.Errorf("error extracting target from message: %v", err)
+	}
+	// make a copy of the targets to keep track of results during retries
+	localTo := make([]string, len(targets))
```
Comment (oryband) [inline http_client.go]: are you sure you didn't mean `make([]string, 0, len(targets))`?

File: http_client.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+			multicastId = gcmResp.MulticastId
+			if err != nil {
+				return gcmResp, fmt.Errorf("error checking GCM results: %v", err)
+			}
+			if doRetry {
+				retryAfter, err := time.ParseDuration(c.getRetryAfter())
+				if err != nil {
+					b.setMin(retryAfter)
+				}
+				localTo = make([]string, len(toRetry))
```
Comment (oryband) [inline http_client.go]: same as previous comment

File: http_client.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+		gcmResp = buildRespForMulticast(targets, *resultsState, multicastId)
+	}
+	return gcmResp, nil
+}
+
+// httpGcmClient implementation to send a message through GCM Http server.
+func (c *httpGcmClient) send(m HttpMessage) (*HttpResponse, error) {
+	bs, err := json.Marshal(m)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling message>%v", err)
```
Comment (oryband) [inline http_client.go]: the message doesn't look right. why use `>`?

File: http_client.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	}
+	debug("sending", string(bs))
+	req, err := http.NewRequest("POST", c.GcmURL, bytes.NewReader(bs))
+	if err != nil {
+		return nil, fmt.Errorf("error creating request>%v", err)
+	}
+	req.Header.Add(http.CanonicalHeaderKey("Content-Type"), "application/json")
+	req.Header.Add(http.CanonicalHeaderKey("Authorization"), fmt.Sprintf("key=%v", c.apiKey))
```
Comment (oryband) [inline http_client.go]: use `strings.Join()` instead of slow `Sprintf`

File: http_client.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+		result, ok := mrs[regId]
+		if !ok {
+			continue
+		}
+		resp.Results[i] = *result
+		if result.MessageId != "" {
+			resp.Success++
+		} else if result.Error != "" {
+			resp.Failure++
+		}
```
Comment (oryband) [inline http_client.go]: this is better off in a `switch`

File: http_client.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+// Transform the recipient in an array of strings if needed.
+func messageTargetAsStringsArray(m HttpMessage) ([]string, error) {
+	if m.RegistrationIds != nil {
+		return m.RegistrationIds, nil
+	} else if m.To != "" {
+		target := []string{m.To}
+		return target, nil
+	}
```
Comment (oryband) [inline http_client.go]: `switch`

File: http_client.go [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+	target := []string{}
+	return target, fmt.Errorf("can't find any valid target field in message.")
```
Comment (oryband) [inline http_client.go]: `errors.New()`, and remove the dot at the end

Comment (oryband) [inline http_client.go]: `return []string{}, errors.New()`

File: http_client.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	doRetry = false
+	toRetry = []string{}
+	for i := 0; i < len(gcmResults); i++ {
+		result := gcmResults[i]
+		regId := recipients[i]
+		resultsState[regId] = &result
+		if result.Error != "" {
+			if retryableErrors[result.Error] {
+				toRetry = append(toRetry, regId)
+				if doRetry == false {
```
Comment (oryband) [inline http_client.go]: `if !doRetry {` ?

File: http_client.go (hunk 2 above)
Comment (bolshoy) [inline http_client.go]: this is google's code and it works, prefer not to touch it.
to your question, i think both do the same in the case of string. You create a preallocated slice with nil elements, he creates an allocated slice with empty strings which are nils. For a big structure slice element, you are right, unnecessary copy is done.

File: http_client.go (hunk 4 above)
Comment (bolshoy) [inline http_client.go]: google's code, I already started to remove these

**Comment by bolshoy** [comment]
@ory mostly google's code here, he's not proficient enough in go. If we decide to completely fork out, I will ask you to do a proper facelift here)

**Comment by Ory** [comment]
ok

**Comment by oryband** [comment]
@bolshoy delete the branch

**Pull request merged**